api_factory/generator/formatter.py

```diff
@@ -0,0 +1,41 @@
+# Copyright 2018 Google LLC
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     https://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+import re
+
+
+def fix_whitespace(code: str) -> str:
+    """Perform basic whitespace post-processing.
+
+    This corrects a couple of formatting issues that Jinja templates
+    may struggle with (particularly blank line count, which is tough to
+    get consistently right when ``if`` or ``for`` are involved).
+
+    Args:
+        code (str): A string of code to be formatted.
+
+    Returns
+        str: Formatted code.
+    """
+    # Remove trailing whitespace from any line.
+    code = re.sub(r'[ ]+\n', '\n', code)
+
+    # Ensure at most two blank lines before top level definitions.
+    code = re.sub(r'\s+\n\n\n(class|def|@)', r'\n\n\n\1', code)
```

Comment:
How bothersome is it to maintain these things at the template level? I think it is good practice to have the templates be as clean as possible.

Maybe we can introduce a logger that emits a developer warning if any of these substitutions are called and have PRs fail on these developer warnings.

Reply:
In general, formatting at the template level is easy, but I have found a couple issues where _this one thing_ is hard.

Basically the problem comes down to `{% for %}` and `{% if %}`. You can use the `{%-` thing in Jinja to do line collapses and such, but it becomes very tricky to do them where a block may or may not be present, or will be present multiple times (which tends to end up with first/last idiosyncrasies).

I have gotten the templates pretty close but there tends to be cases where there are very slight errors in some cases, which these fix.

However, I am also _very_ good with Jinja (having used it for many non-trivial projects over a decade); for a naive template writer, this would be the thing that would cause probably more pain than every other formatting issue combined, and this eliminates it.

Essentially, I agree with you in principle. This case is the exception that proves the rule.

Reply:
For clarity, because I did not explicitly call this out:

> Maybe we can introduce a logger that emits a developer warning if any of these substitutions are called and have PRs fail on these developer warnings.

No, because there are literally a couple cases where they are not avoidable at all, which is why I added this.

Reply:
Ah gotcha, if there are things that are more pain that it's worth as well as being impossible this method is welcome!

Reply:
We should add a template style guide to the docs and enumerate these specific places where enforcing python style through the templates is very hard/impossible as exceptions.